conjure-java-core/src/main/java/com/palantir/conjure/java/types/BeanGenerator.java

```diff
@@ -113,125 +108,67 @@ public static JavaFile generateBeanType(
-                                .collect(ImmutableList.toImmutableList()),
-                        safetyEvaluator);
-
-                List<ClassName> interfacesAsClasses = interfaces.stream()
-                        .map(stageInterface ->
-                                ClassName.get(objectClass.packageName(), objectClass.simpleName(), stageInterface.name))
-                        .collect(Collectors.toList());
-
-                TypeSpec builderInterface = TypeSpec.interfaceBuilder("Builder")
-                        .addModifiers(Modifier.PUBLIC)
-                        .addMethods(interfaces.stream()
-                                .map(stageInterface -> stageInterface.methodSpecs)
-                                .flatMap(List::stream)
-                                .map(method -> method.toBuilder()
-                                        .addAnnotation(Override.class)
-                                        .returns(
-                                                method.name.equals("build")
-                                                        ? method.returnType
-                                                        : ClassName.get(
-                                                                objectClass.packageName(),
-                                                                objectClass.simpleName(),
-                                                                "Builder"))
-                                        .build())
-                                .collect(Collectors.toSet()))
-                        .addSuperinterfaces(interfacesAsClasses.stream()
-                                .map(Primitives::box)
-                                .collect(Collectors.toList()))
-                        .build();
-
-                typeBuilder
-                        .addTypes(interfaces)
-                        .addType(builderInterface)
-                        .addMethod(MethodSpec.methodBuilder("builder")
-                                .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
-                                .returns(interfacesAsClasses.get(0))
-                                .addStatement("return new DefaultBuilder()")
-                                .build())
-                        .addType(BeanBuilderGenerator.generate(
-                                typeMapper,
-                                safetyEvaluator,
-                                objectClass,
-                                builderClass,
-                                typeDef,
-                                typesMap,
-                                options,
-                                Optional.of(ClassName.get(
-                                        objectClass.packageName(), objectClass.simpleName(), builderInterface.name))));
-            }
+        if (hasFields) {
+            addBuilder(typeBuilder, objectClass, fields, typeMapper, safetyEvaluator, typeDef, typesMap, options);
```

Comment:
Motivation:
I don't think we expect to ever use `generateStagedInterfaces` without also generating all of the remaining methods. It seems reasonable to have one method do all of the required generation to add staged builders.
 
```java
typeBuilder
     .addTypes(interfaces)
     .addType(builderInterface)
     .addMethod(MethodSpec.methodBuilder("builder")
     .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
     .returns(interfacesAsClasses.get(0))
     .addStatement("return new DefaultBuilder()")
     .build())
.addType(BeanBuilderGenerator.generate(...));
```

In general I don't like the fact that `addBuilder` has side effects, but given the method is either in this class, or some other tightly coupled class, it seems reasonable.

Reply:
Part of the issue here is that the builder was originally a single concrete `Builder` class enclosed within the bean type, however Staged builders require N builder interfaces, which are also nested within the bean type, not contained in the builder, so `BeanBuilderGenerator.generate` can no longer simply return a type. It could consume the outer typeBuilder instead, or spec out an object which represents N additions (a list of new typespecs, methods, fields, etc in a more pure way than mutating a builder instance).

Reply:
> consume the outer typeBuilder instead

Yep. I think that maps to having the `addBuilder` method here, be in `BeanBuilderGenerator`.

> spec out an object which represents N additions

I imagine this would look something like:
```java
public record GeneratedStatements (
    List<AnnotationSpec> annotations, 
    List<InterfaceSpec> interfaces, 
    // etc.
    ...) {}
```

then in our caller's code:
```java
stmts = BeanBuilderGenerator.generateStagedBuilder(...)
typeBuilder.addMethods(stmts.methods).addInterfaces(stmts.interfaces)...
```

If something like that is what you have in mind, I think we'd have to be careful about the insertion order of statements in `BeanBuilderGenerator`.

But I think it's a little annoying that we need to reflect the ordering of the statements in the generated code, by the call order of `typeBuilder.addXYZ`, and can't pass in our own ordering.

Reply:
Ah I guess we could have some union type `Union[AnnotationSpec, InterfaceSpec, ...]` which would allow me to just store one list instead of many. Still think passing in the `builder` is more straightforward.

> pass in our own ordering

thinking a bit more about what this might look like, maybe we could tag each poet spec with some string/enum/class, and provide a list of those string/enum/class as an ordering.

maybe:
```java
public record Statement<T> (T type, Union[AnnotationSpec, InterfaceSpec, ...] spec) {
}

public record GeneratedStatements<T> (
    List<Statement<T>> statements, 
    // maybe some other metadata
    ...) {}

public <StatementT> TypeBuilder.Builder generate(GeneratedStatements<StatementT> stmts, List<StatementT> order) {
    // stable sort stmts.statements according to the type's index in `order` 
    // for stmt : sortedStmts: add the statement to builder (unsure exactly how. maybe use a visitor) 
}
```

then in a generator class:
```java
// allow each "generator" class to tag statements with their own values
class MyGeneratedStatementType extends GeneratedStatementType {
    MY_ANNOTATION, 
    MY_INTERFACE,
    SOME_MODIFIER,
    //...
}

List<Statement<MyGeneratedStatementType>> stmts = new ArrayList<>();
stmts.add(Statement(MY_ANNOTATION, annotationSpec));
stmts.add(Statement(MY_INTERFACE, interfaceSpec))
generate(typeBuilder, stmts, List.of(MY_ANNOTATION, SOME_MODIFIER, MY_INTERFACE, ...)).build();
```

There are likely better ways, but all of this seems like overkill.